### ISSUE TITLE
Race condition when sharing Engine instance between multiple threads

### ISSUE THREAD

**Issue by jaroslawr** [oracle/graal#4176]
**Describe the issue**

Sharing an Engine instance between multiple threads causes multiple different GuestToHostCodeCache instances to come into existence and ultimately an AssertionError if JVM is running with assertions enabled.

**Steps to reproduce the issue**

With Graal 21.3.0+Java17 this test reproduces the issue:

```java
package com.mycompany.app;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.atomic.AtomicInteger;

import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.Engine;
import org.graalvm.polyglot.HostAccess;
import org.graalvm.polyglot.Source;
import org.graalvm.polyglot.Value;
import org.junit.Test;

public class GraalTest {
    public static class Counter {
        private final AtomicInteger counter;

        public Counter() {
            this.counter = new AtomicInteger(0);
        }

        @HostAccess.Export
        public void increment() {
            counter.incrementAndGet();
        }
    }

    @Test
    public void testGraalEngineSharing() {
        Counter counter = new Counter();

        ExecutorService executorService = Executors.newFixedThreadPool(10);
        Engine engine = Engine.newBuilder().build();
        List<Future> futures = new ArrayList<>();

        for (int i = 0; i < 10; i++) {
            final String script = "(function(counter) { for (var i = 0; i < 1000000; i++) { counter.increment(); }; return " + i + "; });";
            futures.add(executorService.submit(new Callable<String>() {
                public String call() {
                    try {
                        System.out.println("START");

                        Context context = Context.newBuilder("js")
                                .engine(engine)
                                .build();

                        Source source = Source.create("js", script);

                        Value fn = context.eval(source);
                        fn.execute(counter);

                    } catch (Exception e) {
                        System.out.println(e);
                        e.printStackTrace();
                    }

                    return "x";
                }
            }));
        }

        for (Future<String> future : futures) {
            try {
                future.get();
            } catch(Exception e) {
                System.out.println("Exception during get");
                System.out.println(e);
                e.printStackTrace();
            }
        }
    }
}
```

```
org.graalvm.polyglot.PolyglotException: java.lang.AssertionError
        at org.graalvm.truffle/com.oracle.truffle.host.HostExecuteNode.doInvoke(HostExecuteNode.java:871)
        at org.graalvm.truffle/com.oracle.truffle.host.HostExecuteNode.doFixed(HostExecuteNode.java:137)
        at org.graalvm.truffle/com.oracle.truffle.host.HostExecuteNodeGen.executeAndSpecialize(HostExecuteNodeGen.java:143)
        at org.graalvm.truffle/com.oracle.truffle.host.HostExecuteNodeGen.execute(HostExecuteNodeGen.java:96)
        at org.graalvm.truffle/com.oracle.truffle.host.HostObject.invokeMember(HostObject.java:451)
        at org.graalvm.truffle/com.oracle.truffle.host.HostObjectGen$InteropLibraryExports$Cached.invokeMemberNode_AndSpecialize(HostObjectGen.java:2695)
        at org.graalvm.truffle/com.oracle.truffle.host.HostObjectGen$InteropLibraryExports$Cached.invokeMember(HostObjectGen.java:2677)
        at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibrary$Asserts.invokeMember(InteropLibrary.java:3450)
        at org.graalvm.truffle/com.oracle.truffle.api.interop.InteropLibraryGen$CachedDispatch.invokeMember(InteropLibraryGen.java:8138)
        at com.oracle.truffle.js.nodes.function.JSFunctionCallNode$ForeignInvokeNode.executeCall(JSFunctionCallNode.java:1475)
        at com.oracle.truffle.js.nodes.function.JSFunctionCallNode.executeAndSpecialize(JSFunctionCallNode.java:303)
        at com.oracle.truffle.js.nodes.function.JSFunctionCallNode.executeCall(JSFunctionCallNode.java:249)
        at com.oracle.truffle.js.nodes.function.JSFunctionCallNode$InvokeNode.execute(JSFunctionCallNode.java:722)
        at com.oracle.truffle.js.nodes.JavaScriptNode.executeVoid(JavaScriptNode.java:191)
        at com.oracle.truffle.js.nodes.control.AbstractBlockNode.executeVoid(AbstractBlockNode.java:80)
        at com.oracle.truffle.js.nodes.control.AbstractBlockNode.executeVoid(AbstractBlockNode.java:55)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedBlockNode.executeVoid(OptimizedBlockNode.java:120)
        at com.oracle.truffle.js.nodes.control.AbstractBlockNode.executeVoid(AbstractBlockNode.java:70)
        at com.oracle.truffle.js.nodes.control.AbstractRepeatingNode.executeBody(AbstractRepeatingNode.java:67)
        at com.oracle.truffle.js.nodes.control.WhileNode$WhileDoRepeatingNode.executeRepeating(WhileNode.java:231)
        at org.graalvm.truffle/com.oracle.truffle.api.nodes.RepeatingNode.executeRepeatingWithValue(RepeatingNode.java:112)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedOSRLoopNode.profilingLoop(OptimizedOSRLoopNode.java:158)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedOSRLoopNode.execute(OptimizedOSRLoopNode.java:109)
        at com.oracle.truffle.js.nodes.control.WhileNode.executeVoid(WhileNode.java:178)
        at com.oracle.truffle.js.nodes.control.AbstractBlockNode.executeVoid(AbstractBlockNode.java:80)
        at com.oracle.truffle.js.nodes.control.AbstractBlockNode.executeVoid(AbstractBlockNode.java:55)
        at jdk.internal.vm.compiler/org.graalvm.compiler.truffle.runtime.OptimizedBlockNode.executeGeneric(OptimizedBlockNode.java:79)
        at com.oracle.truffle.js.nodes.control.AbstractBlockNode.execute(AbstractBlockNode.java:75)
        at com.oracle.truffle.js.nodes.binary.DualNode.execute(DualNode.java:125)
        at com.oracle.truffle.js.nodes.function.FunctionBodyNode.execute(FunctionBodyNode.java:73)
        at com.oracle.truffle.js.nodes.function.FunctionRootNode.executeInRealm(FunctionRootNode.java:150)
        at com.oracle.truffle.js.runtime.JavaScriptRealmBoundaryRootNode.execute(JavaScriptRealmBoundaryRootNode.java:93)
        at <js> :anonymous(Unknown)
        at org.graalvm.sdk/org.graalvm.polyglot.Value.execute(Value.java:841)
        at com.mycompany.app.AppTest$1.call(AppTest.java:54)
        at com.mycompany.app.AppTest$1.call(AppTest.java:42)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by rodrigar-mx** [comment]
Hi @jaroslawr. Thanks for contributing to graalvm. This issue has been reported internally. You will receive a response accordingly.

**Comment by woess** [comment]
Thank you for reporting this and providing a reproducer! Fixed in fc5d6bef63c71faad0caf2e41af92439e1098120.

**Comment by jaroslawr** [comment]
@woess Thanks for the fix! I see its part of vm-22.0.0.2 tag, but maven central only has graal 22.0.0. Will 22.0.0.2 eventually make it to maven central?

**Comment by woess** [comment]
@jaroslawr it should be part of 22.0.0.1 and released soon:tm:
